File: snippets/rust/grant/grant-quote.rs
//@! start chunk 1 | title=Import dependencies
use open_payments::client::api::UnauthenticatedResources;
use open_payments::client::AuthenticatedResources;
use open_payments::snippets::utils::{create_authenticated_client, get_env_var, load_env};
use open_payments::types::{
    auth::{AccessItem, AccessTokenRequest, GrantRequest, QuoteAction},
    GrantResponse,
};
//@! end chunk 1

#[tokio::main]
async fn main() -> open_payments::client::Result<()> {
    load_env()?;

    //@! start chunk 2 | title=Initialize Open Payments client
    // Authenticated client can be also used for unauthenticated resources
    let client = create_authenticated_client()?;
    //@! end chunk 2

    //@! start chunk 3 | title=Get wallet address information
    let wallet_address_url = get_env_var("WALLET_ADDRESS_URL")?;
    let wallet_address = client.wallet_address().get(&wallet_address_url).await?;
    //@! end chunk 3

    //@! start chunk 4 | title=Request quote grant
    let grant_request = GrantRequest::new(
        AccessTokenRequest {
            access: vec![AccessItem::Quote {
                actions: vec![QuoteAction::Create, QuoteAction::Read, QuoteAction::ReadAll],
            }],
        },
        None,
    );

    println!(
        "Grant request JSON: {}",
        serde_json::to_string_pretty(&grant_request)?
    );

    let response = client
        .grant()
        .request(&wallet_address.auth_server, &grant_request)
        .await?;
    //@! end chunk 4

    //@! start chunk 5 | title=Output
    match response {
        GrantResponse::WithToken { access_token, .. } => {
            println!("Received access token: {:#?}", access_token.value);
            println!(
                "Received access token manage URL: {:#?}",
                access_token.manage
            );
        }
        GrantResponse::WithInteraction { .. } => {
            unreachable!("Interaction not required for quotes");
        }
    }
    //@! end chunk 5

    Ok(())
}

File: snippets/php/grant/grant-quote.php
<?php

//@! start chunk 1 | title=Get wallet address information
$wallet = $opClient->walletAddress()->get([
    'url' => $config->getWalletAddressUrl()
]);
//@! end chunk 1

//@! start chunk 2 | title=Request quote grant
$grant = $opClient->grant()->request(
    [
        'url' => $wallet->authServer
    ],
    [
        'access_token' => [
            'access' => [
                [
                    'type' => 'quote',
                    'actions' => ['create', 'read', 'read-all']
                ]
            ]
        ],
        'client' => $config->getWalletAddressUrl()
    ]
);
//@! end chunk 2

//@! start chunk 3 | title=Check grant state
 if ($grant instanceof \OpenPayments\Models\PendingGrant) {
    throw new \Error('Expected non-interactive grant');
}
//@! end chunk 3

//@! start chunk 4 | title=Output
echo 'QUOTE_ACCESS_TOKEN: ' . $grant->access_token->value . PHP_EOL;
echo 'QUOTE_ACCESS_TOKEN_MANAGE_URL: ' . $grant->access_token->manage . PHP_EOL;

echo 'GRANT OBJECT: ' . PHP_EOL . print_r($grant, true);
//@! end chunk 4
